Source/Diagnostics/ReducedDiags/FieldEnergy.cpp

```diff
@@ -177,20 +180,35 @@ FieldEnergy::ComputeNorm2RZ(const amrex::MultiFab& field, const int lev)
         amrex::Box tb = convert(tilebox, field.ixType().toIntVect());
 
         // Lower corner of tile box physical domain
-        const std::array<amrex::Real, 3>& xyzmin = WarpX::LowerCorner(tilebox, lev, 0._rt);
+        const std::array<amrex::Real, 3>& xyzmin = warpx.LowerCorner(tilebox, lev, 0._rt);
         const Dim3 lo = lbound(tilebox);
-        const Real rmin = xyzmin[0] + (tb.ixType().nodeCentered(0) ? 0. : 0.5*dx[0]);
+        const Dim3 hi = ubound(tilebox);
+        const Real rmin = xyzmin[0] + (tb.ixType().nodeCentered(0) ? 0._rt : 0.5_rt*dr);
         const int irmin = lo.x;
+        const int irmax = hi.x;
 
         int const ncomp = field.nComp();
 
-        const amrex::Box & box = enclosedCells(mfi.nodaltilebox());
-        reduce_ops.eval(box, ncomp, reduce_data,
+        for (int idir=0 ; idir < AMREX_SPACEDIM ; idir++) {
+            if (warpx.field_boundary_hi[idir] == FieldBoundaryType::Periodic) {
+                // For periodic boundaries, do not include the data in the nodes
+                // on the upper edge of the domain
+                tb.enclosedCells(idir);
+            }
+        }
+
+        reduce_ops.eval(tb, ncomp, reduce_data,
             [=] AMREX_GPU_DEVICE (int i, int j, int k, int n) -> ReduceTuple
             {
                 const amrex::Real r = rmin + (i - irmin)*dr;
+                amrex::Real volume_factor = r;
```

Comment:
Hi @dpgrote , let's calculate this factor before the `eval` and capture the variable `volume_factor` automatically in the lambda.